Mathlib/AlgebraicGeometry/Morphisms/Affine.lean

```diff
@@ -0,0 +1,237 @@
+- `AlgebraicGeometry.isAffineOpen_of_isAffineOpen_basicOpen`:
+  If `s` is a spanning set of `Γ(X, U)`, such that each `X.basicOpen i` is affine,
+  then `U` is also affine.
+- `AlgebraicGeometry.isAffineHom_isLocalAtTarget`: Affine morphisms are local at the target.
+- `AlgebraicGeometry.isAffineHom_stableUnderBaseChange`:
+  Affine morphisms are stable under base change.
+- `AlgebraicGeometry.isAffineHom_iff_isAffine`:
+
+
+
+-/
+
+universe v u
+
+open CategoryTheory TopologicalSpace Opposite
+
+namespace AlgebraicGeometry
+
+variable {X Y Z : Scheme.{u}} (f : X ⟶ Y) (g : Y ⟶ Z)
+
+/-- A morphism of schemes `X ⟶ Y` is affine if
+the preimage of any affine open subset of `Y` is affine. -/
+@[mk_iff]
+class IsAffineHom {X Y : Scheme} (f : X ⟶ Y) : Prop where
+  isAffine_preimage : ∀ U : Opens Y, IsAffineOpen U → IsAffineOpen (f ⁻¹ᵁ U)
+
+/-- The preimage of an affine open as an `Scheme.affine_opens`. -/
+@[simps]
+def affinePreimage {X Y : Scheme} (f : X ⟶ Y) [IsAffineHom f] (U : Y.affineOpens) :
+    X.affineOpens :=
+  ⟨f ⁻¹ᵁ U.1, IsAffineHom.isAffine_preimage _ U.prop⟩
+
+instance (priority := 900) [IsIso f] : IsAffineHom f :=
+  ⟨fun _ hU ↦ hU.map_isIso f⟩
+
+instance (priority := 900) [IsAffineHom f] : QuasiCompact f :=
+  (quasiCompact_iff_forall_affine f).mpr
+    (fun U hU ↦ (IsAffineHom.isAffine_preimage U hU).isCompact)
+
+instance [IsAffineHom f] [IsAffineHom g] : IsAffineHom (f ≫ g) := by
+  constructor
+  intros U hU
+  rw [Scheme.comp_val_base, Opens.map_comp_obj]
+  apply IsAffineHom.isAffine_preimage
+  apply IsAffineHom.isAffine_preimage
+  exact hU
+
+instance : MorphismProperty.IsMultiplicative @IsAffineHom where
+  id_mem := inferInstance
+  comp_mem _ _ _ _ := inferInstance
+
+/-- The `AffineTargetMorphismProperty` corresponding to affine morphisms. -/
+def IsAffineHom.affineProperty : AffineTargetMorphismProperty :=
+  fun X _ _ _  ↦ IsAffine X
+
+@[simp] lemma IsAffineHom.affineProperty_toProperty :
+    AffineTargetMorphismProperty.toProperty IsAffineHom.affineProperty f ↔
+      IsAffine Y ∧ IsAffine X := by
+  delta AffineTargetMorphismProperty.toProperty IsAffineHom.affineProperty; simp
+
+lemma isAffineHom_iff_affineProperty :
+    IsAffineHom f ↔ targetAffineLocally IsAffineHom.affineProperty f :=
+  (isAffineHom_iff f).trans ⟨fun H U ↦ H U U.prop, fun H U hU ↦ H ⟨U, hU⟩⟩
+
+lemma isAffineHom_eq_affineProperty :
+    @IsAffineHom = targetAffineLocally IsAffineHom.affineProperty := by
+  ext; exact isAffineHom_iff_affineProperty _
+
+instance {X : Scheme} (r : Γ(X, ⊤)) :
+    IsAffineHom (Scheme.ιOpens (X.basicOpen r)) := by
+  constructor
+  intros U hU
+  fapply (Scheme.Hom.isAffineOpen_iff_of_isOpenImmersion (Scheme.ιOpens _)).mp
+  convert hU.basicOpen (X.presheaf.map (homOfLE le_top).op r)
+  rw [X.basicOpen_res]
+  ext1
+  refine Set.image_preimage_eq_inter_range.trans ?_
+  erw [Subtype.range_coe]
+  rfl
+
+lemma iSup_basicOpen_eq_top_of_span_eq_top {X : Scheme} (s : Set (X.presheaf.obj (op ⊤)))
```

Comment:
This PR looks very good to me, apart from the fact that this lemma (and the next) may be moved to a slightly better file.